### PR TITLE
Subpackages

### PR DESCRIPTION
**Pull request by olivierwilkinson**
### TLDR;

The SDK is large so it would be nice to only have to install the parts we need for a particular challenge. Splitting it up is not urgent, and it is not really essential either. This PR shows how to split up the SDK into subpackages by reorganising the directories and adding setup.py files, but without any breaking changes or changes to the application code.

### Motivation

There has been recent discussion about how we manage code splitting of the SDK, in particular in relation to using parts of the SDK without needing to download large files found in processing and robotics.

This would be beneficial when version locking dependencies in challenges. Challenges that only need to use the `pma` components would be able to specify `pitop.pma: vX.X.X` and we could pull just that fragment of the SDK.

This is not _essential_ however as when creating a virtualenv we can use the `--system-site-packages` flag to use the host's packages if they match the specified version. This means only the packages that are necessary are installed. This does not replace splitting up the sdk but it does make the need non-urgent.

### This Solution

The current structure of the SDK is `pitop > package`, for example the `pitop.pma` can be found at `pitop > pma`. `pitop` is being used as a namespace, by  flattening out our directory structure and leveraging [namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/) we can split out each directory into their own package without any breaking changes.

The end result of doing things this way is that:
- subpackages can be installed directly by installing their package name prefixed by `pitop`, our namespace. 
```shell
pip3 install pitop.pma
```
- importing from subpackages stays the same, meaning the code is identical within the subpackages and there are no breaking changes.
```python3
from pitop.pma import Button
```
- interdepencies between subpackages becomes clearer as each subpackage must declare it's own dependencies. This should allow us to improve isolation over time. For example the `pma` subpackage depends on the `core` subpackage, this means `pma/setup.py` must declare `pitop.core` in it's `__requires__` array:
```python3
__requires__ = [
    ...
    "pitop.core==vX.X.X"
    ...
]
```

This PR shows how to structure the packages to achieve this:
- It flattens the packages so they are no longer nested in pitop 
- each package has it's own setup.py file and declares it's own project and deps
- The CURRENT_VERSION environment variable is used to set every packages version, allowing subpackages to be published in step; this means knowing which packages work together is trivial.
- The packages themselves are organised so they are in the `pitop` namespace.
- I left some packages nested in the `pitop`package to illustrate that we can still keep stuff there in the same way we used to.
- the `pitop` package declares all other subpackages as it's deps and keeps the same top-level exports as before.

### Next Steps

This PR is **not** ready to be merged:
- ~I tried to make sure each subpackage has the correct deps but didn't double check~
- ~I do not currently have my pitop with me so I can't correctly install every package so there might be errors there~
- ~I ignored debian packaging other than to think about how that would work. (only need to keep publishing the current deb package, there are flags to [specify which directories to use](https://manpages.debian.org/testing/dh-python/pybuild.1.en.html#DIRECTORIES) when running pybuild)~
- ~I did not update CI~
- ~I completely ignored docs.~

### Testing

1. Test upgrading pi-topOS

- Download PR artifacts into your pi-top.
- Install them by running `sudo apt install ./*.deb` in the folder where the deb files are.
- Process should be able to complete without issues/warnings.
- Make sure everything still works (eg: run a python script that uses the miniscreen)

2. Test standalone packages (deb files)

For this section, I suggest using Raspberry Pi OS. Also, this will be focused in using the Miniscreen package (`python3-pitop-miniscreen`) but the instructions can be extended to any of the other new subpackages.

2.1. Prepare the OS

```
sudo apt update
sudo apt install -y pi-top-os-unstable-apt-source
```

2.2. Install the miniscreen subpackage

Try to install the `python3-pitop-miniscreen` package. Since the packages are not on any mirror, `apt` will fail to find the dependencies:

```
pi@raspberrypi:~ $ sudo apt install ./python3-pitop-miniscreen_0.24.3-2~25.gbpf29847_all.deb 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Note, selecting 'python3-pitop-miniscreen' instead of './python3-pitop-miniscreen_0.24.3-2~25.gbpf29847_all.deb'
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 python3-pitop-miniscreen : Depends: python3-pitop-common but it is not installable
                            Depends: python3-pitop-core but it is not installable
E: Unable to correct problems, you have held broken packages.
```

So we need to provide the path to the deb that provide those dependencies explicitly:
```
sudo apt install ./python3-pitop-miniscreen_0.24.3-2~25.gbpf29847_all.deb  ./python3-pitop-common_0.24.3-2~25.gbpf29847_all.deb  ./python3-pitop-core_0.24.3-2~25.gbpf29847_all.deb
```

2.3. Install pi-topd

Install `pi-topd` to make sure we can use the miniscreen. Download the artifacts from this PR https://github.com/pi-top/pi-topd/pull/169 and install them.

2.4. Play with the miniscreen

Use the miniscreen using the python interpreter:

```
pi@raspberrypi:~ $ python3
Python 3.9.2 (default, Mar 12 2021, 04:06:34) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pitop.miniscreen
>>> m = pitop.miniscreen.Miniscreen()
>>> m.display_text("hey!")
```

3. Using PyPI

3.1. Clean previous steps

Remove the packages we installed in step (2):
```
sudo apt purge python3-pitop-miniscreen python3-pitop-common python3-pitop-core
sudo apt autoremove 
```

This will remove `pi-topd` too, since we're removing one of its dependencies, but lets not worry about that for now.

3.2. Install `pitop.miniscreen`

The new subpackages are available on test PyPI. To install `pitop.miniscreen`, run:

```
python3 -m pip install --index-url https://test.pypi.org/simple/ pitop.miniscreen
```

3.3. Re-install pi-topd

Even though the `pitop.common` package is installed via PyPI, we still need to provide the deb when installing `pi-topd` via apt:

```
sudo apt install ./pi-topd*.deb ./python3-pitop-common_0.24.3-2~25.gbpf29847_all.deb 
```

3.4. Install other dependencies

To use the miniscreen, we need to install the fonts used by it. These are automatically installed when using the debs but here we need to do it manually. Also, the `libatlas-base-dev` package is required by opencv.

```
sudo apt install fonts-roboto ttf-bitstream-vera libatlas-base-dev -y
```

3.5. Test

TODO: Found some issues with opencv and numpy by this step! OpenCV doesn't like the pre-installed version of numpy:
```
>>> import cv2
RuntimeError: module compiled against API version 0xf but this version of numpy is 0xd
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/pi/.local/lib/python3.9/site-packages/cv2/__init__.py", line 8, in <module>
    from .cv2 import *
ImportError: numpy.core.multiarray failed to import
```
Numpy is installed out of the box with Raspberry Pi OS. Manually upgrading it fixes this issue.

### CODE REVIEW

**Comment by m-roberts** [comment]
Would close #401

**Comment by m-roberts** [comment]
Understandably, given refactoring on main branch, this PR has got some conflicts.

**Comment by olivierwilkinson** [comment]
Since almost every change in this PR is considered by git to be a rename then we can rebase this branch onto main, then the most recent version of the file gets renamed. Any new files would need to be manually moved while rebasing but that is pretty simple since they just needs to be put in the corresponding package folder. That's what I did before I went off and it only took about 5 mins so don't stress about this branch! 👍

**Comment by m-roberts** [comment]
So from my understanding of namespace packages in Python (using the link you provided), the `__init__.py` file has to be stripped of importing actual objects, modules, etc.
Implicit PEP 420 namespaces mean it must be omitted, and the only other namespace alternatives require a file to be _only_ `__path__ = __import__('pkgutil').extend_path(__path__, __name__)` or `__import__('pkg_resources').declare_namespace(__name__)`.

This means that this implementation breaks backwards compatibility for `from pitop import *`.

Perhaps we _could_ release these subpackages under a new `pi_top` namespace, and keep `pitop` for high-level importing/legacy support?

**Comment by olivierwilkinson** [comment]
> So from my understanding of namespace packages in Python (using the link you provided), the `__init__.py` file has to be stripped of importing actual objects, modules, etc. Implicit PEP 420 namespaces mean it must be omitted, and the only other namespace alternatives require a file to be _only_ `__path__ = __import__('pkgutil').extend_path(__path__, __name__)` or `__import__('pkg_resources').declare_namespace(__name__)`.
> 
> This means that this implementation breaks backwards compatibility for `from pitop import *`.
> 
> Perhaps we _could_ release these subpackages under a new `pi_top` namespace, and keep `pitop` for high-level importing/legacy support?

So in our case it's fine to extend the path in our pitop packages `__init__.py` file using `__path__ = __import__('pkgutil').extend_path(__path__, __name__)`. This is because while it's true that `pkgutil` requires every `__init__.py` at the namespace level to be identical we actually only have the one that lives in `/packages/pitop`. We will also only ever have the one that lives in `/packages/pitop` since modifying the namespace doesn't make sense in the `pitop.common`: what lives at the `pitop` level is not the purview of that package.

I'll add a commit which extends the path when `pitop` is installed and if you want we can run through it together on a call 👍

**Comment by m-roberts** [comment]
Let's have that chat soon - @jcapona and I think my suggestion may be easier to implement and maintain, but we want to confirm.

For reference, [here](https://github.com/pypa/sample-namespace-packages) is an GitHub repo with the 3 types of namespace packages in Python.

**Comment by olivierwilkinson** [comment]
> Let's have that chat soon - @jcapona and I think my suggestion may be easier to implement and maintain, but we want to confirm.
> 
> 
> 
> For reference, [here](https://github.com/pypa/sample-namespace-packages) is an GitHub repo with the 3 types of namespace packages in Python.

Sure, let's chat it through. It would be good to get on the same page! I don't really follow how it's easier to implement or maintain as it stands.

File: packages/keyboard/setup.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+]
+
+__keywords__ = [
+    "pi-top",
+    "raspberrypi",
+    "battery",
+]
+
+__requires__ = [
+    f"pitop.common=={__version__}",
```
Comment (olivierwilkinson) [inline packages/keyboard/setup.py:44]: I think this needs to include XLib, the vendored pynput package imports it:
```
Traceback (most recent call last):
  File "/home/pi/subpackages/test.py", line 3, in <module>
    button = KeyboardButton("a")
  File "/home/pi/subpackages/.virtualenv/lib/python3.9/site-packages/pitop/keyboard/keyboard_button.py", line 18, in __init__
    from .vendor.pynput.keyboard import Listener
  File "/home/pi/subpackages/.virtualenv/lib/python3.9/site-packages/pitop/keyboard/vendor/pynput/__init__.py", line 22, in <module>
    from . import keyboard, mouse
  File "/home/pi/subpackages/.virtualenv/lib/python3.9/site-packages/pitop/keyboard/vendor/pynput/keyboard/__init__.py", line 48, in <module>
    from ._xorg import Controller, Key, KeyCode, Listener
  File "/home/pi/subpackages/.virtualenv/lib/python3.9/site-packages/pitop/keyboard/vendor/pynput/keyboard/_xorg.py", line 28, in <module>
    import Xlib.display
ModuleNotFoundError: No module named 'Xlib'
```

File: packages/processing/setup.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+]
+
+__requires__ = [
+    f"pitop.common=={__version__}",
+    f"pitop.pma=={__version__}",
+    f"pitop.core=={__version__}",
+    "imutils>=0.5.4,<1.0.0",
+    "matplotlib>=3.3.4,<4.0",
+    "numpy>=1.19.5,<2.0",
+    "onnxruntime>=1.8.1,<2.0",
```
[the remark this answers is not in the document]
Comment (jcapona) [reply packages/processing/setup.py:50]: This happens because pypi/testpypi don't provide a `onnxruntime` wheel for arm... that's why we compile and distribute it ourselves as a debian package. 
We should probably make that wheel available through a pypi-repository via PackageCloud and update any installation instruction to make sure users add it as an extra-index to their pip configuration.

File: packages/keyboard/setup.py (hunk 1 above)
Comment (jcapona) [reply packages/keyboard/setup.py:45]: added `xlib` as a dependency to the `keyboard`  subpackage

File: packages/system/setup.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+__keywords__ = [
+    "pi-top",
+    "raspberrypi",
+    "battery",
+]
+
+__requires__ = [
+    f"pitop.common=={__version__}",
+    f"pitop.core=={__version__}",
```
[the remark this answers is not in the document]
Comment (olivierwilkinson) [reply packages/system/setup.py:45]: Explicitly depending on `pitop.battery` and `pitop.miniscreen` causes a circular dependency, so we must make these soft dependencies. For now I think we should leave the error message as is to avoid changing the behaviour of any packages in this PR.

**Review by jcapona — APPROVED** [review]
